### DISCUSSION TITLE
Using Weights & Biases with AllenNLP

### DISCUSSION THREAD

**Discussion by epwalsh** [allenai/allennlp#5069] [Guides]
![image](https://user-images.githubusercontent.com/8812459/112694532-cb4f7c80-8e3f-11eb-9ab0-8c36050408db.png)

This post will walk you through an example of logging experiments to [Weights & Biases](https://wandb.ai) with AllenNLP via the [`WandBCallback`](https://docs.allennlp.org/main/api/training/callbacks/wandb/#wandbcallback). It assumes you're already familiar with AllenNLP (if not, see [Getting Started Using the Library](https://docs.allennlp.org/main/#getting-started-using-the-library)) and have both `allennlp` and `allennlp-models` version 2.2.0 or greater installed.

You'll also need a W&B account. If you don't have one yet, it's free to make one and you can [**sign up here**](https://wandb.ai/login?signup=true) through GitHub or Google.

## Setup

Let's get set up with a small experiment first. Run the following to create a new directory and download the data you'll need:

```bash
mkdir wandb_test_run
cd wandb_test_run
wget https://raw.githubusercontent.com/epwalsh/nlp-models/master/data/greetings.tar.gz
tar -xzvf greetings.tar.gz
```

Then create an AllenNLP experiment configuration file called `config.jsonnet` and the copy over the following contents (I've also posted it [as a Gist](https://gist.github.com/epwalsh/e7bd19750f6be8d6e23d6a5e371ebb21)):

<details><summary><code>config.jsonnet</code> 👇</summary><br/>
    
    local target_namespace = "target_tokens";
    
    {
        "dataset_reader": {
            "target_namespace": target_namespace,
            "type": "copynet_seq2seq",
            "source_token_indexers": {
                "tokens": {
                    "type": "single_id",
                    "namespace": "source_tokens"
                },
            }
        },
        "vocabulary": {
            "min_count": {
                "source_tokens": 4,
                "target_tokens": 4
            }
        },
        "train_data_path": "train.tsv",
        "validation_data_path": "validation.tsv",
        "model": {
            "type": "copynet_seq2seq",
            "source_embedder": {
                "token_embedders": {
                    "tokens": {
                        "type": "embedding",
                        "vocab_namespace": "source_tokens",
                        "embedding_dim": 25,
                        "trainable": true
                    },
                }
            },
            "encoder": {
                "type": "lstm",
                "input_size": 25,
                "hidden_size": 50,
                "num_layers": 2,
                "dropout": 0,
                "bidirectional": true
            },
            "attention": {
                "type": "bilinear",
                "vector_dim": 100,
                "matrix_dim": 100
            },
            "target_embedding_dim": 10,
            "beam_size": 3,
            "max_decoding_steps": 20,
        },
        "data_loader": {
            "batch_sampler": {
                "type": "bucket",
                "batch_size" : 32,
            }
        },
        "trainer": {
            "optimizer": {
                "type": "sgd",
                "lr": 0.1
            },
            "num_epochs": 2,
            "callbacks": [
                {
                    "type": "wandb",
                    "summary_interval": 1,
                    "batch_size_interval": 1,
                    "should_log_learning_rate": true,
                    "should_log_parameter_statistics": true,
                    "project": "allennlp-testing",
                },
            ],
        }
    }

</details>

Lastly, head over to the W&B dashboard and create a new project called "allennlp-testing". Then create an API key from your [settings page](https://wandb.ai/settings) and set it to the environment variable `WANDB_API_KEY`.

## Running the experiment

After you've completed all the setup steps, just run

```bash
allennlp train config.jsonnet -s run1
```

Once the first epoch of training starts, you'll see a new run going in your W&B project with a funky name like "fiery-butterfly-12" 🔥🦋⓵⓶. The experiment should only take a minute or two to complete, after which the metrics in W&B will be fully populated. Click on the run in W&B to explore.

## Exploring what's logged to W&B

AllenNLP logs a lot of information to W&B, some of which you can disable to save on storage like gradient and parameter statistics. Let's walk through everything that's logged step-by-step.

### Overview

First, if you click on the "Overview" tab on the left side panel, you'll see a lot of useful information about the operating system, Python environment, hardware, etc.

![image](https://user-images.githubusercontent.com/8812459/112695916-1cf90680-8e42-11eb-925b-ba8f123c6607.png)

Keep scrolling down and you'll also see a "Config" section, which actually mirrors the AllenNLP configuration file we used to define this experiment.

![image](https://user-images.githubusercontent.com/8812459/112696126-837e2480-8e42-11eb-96e4-315fb38d860d.png)

Cool! Let's keep moving.

### Charts

My favorite tab is probably "Charts". This tab is pretty self-explanatory. It contains charts for all of the batch and epoch metrics that AllenNLP tracks during training and validation.

![image](https://user-images.githubusercontent.com/8812459/112696253-b32d2c80-8e42-11eb-82ea-e87adf24e9b4.png)

You'll notice that the default x-axis variable is just "Step", which corresponds to the training batch number. But when viewing epoch metrics it usually makes more sense to change this to "epoch".

![image](https://user-images.githubusercontent.com/8812459/112696797-baa10580-8e43-11eb-9a6e-2162ee4b5271.png)

Another neat thing about the Charts tab is that you can reorganize it however you'd like! For example, you could drag train and validation **epoch** metrics together into the same section, and leave **batch** metrics on their own.

### Files

Lastly I'll take about the "Files" tab. W&B uploads a few files by default, including:

- `output.log`: the logs captured by W&B,
- `config.yaml`: the W&B config (pretty much the same as the AllenNLP config, but in YAML format),
- `requirements.txt`: a snapshot of the Python environment.

AllenNLP also adds:

- `config.json`: the AllenNLP experiment configuration file in JSON form,
- `out.log`: all of the logs produced from AllenNLP.

It's worth noting that `out.log` will usually contain more lines than `output.log` since W&B can only capture logs from when it's initialized at the start of training, whereas `out.log` will contain ALL of the logs back to when the `allennlp command` is invoked.

![image](https://user-images.githubusercontent.com/8812459/112696556-4e260680-8e43-11eb-9b02-b2d4b74bdecd.png)

---

That's it! This is a brand new feature in AllenNLP so please let us know if you find any bugs or have suggestions for improvements!

**Comment by dirkgr** [comment]
This is great! It'll really pay off when you run that experiment many times and you don't have to do the setup every time.

**Reply by epwalsh** [reply]
Yes for sure. W&B will automatically create new run each time `allennlp train` is invoked. And the sorting+filtering of runs feature is really nice.

![image](https://user-images.githubusercontent.com/8812459/112699395-f2f71280-8e48-11eb-85ad-f33cbd42df72.png)

You can sort by arbitrary things like a certain field in your configuration files.

![image](https://user-images.githubusercontent.com/8812459/112699490-2afe5580-8e49-11eb-8f38-f88ef39e27eb.png)

**Comment by epwalsh** [comment]
Just merged some additional improvements! ➡️ https://github.com/allenai/allennlp/pull/5114

These will go out with the 2.3 release coming soon!
